File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteProjectUserRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.krystianwsul.checkme.firebase.UserData;
import com.krystianwsul.checkme.firebase.json.UserJson;
import com.krystianwsul.checkme.utils.Utils;

import junit.framework.Assert;

import java.util.Map;

public class RemoteProjectUserRecord extends RemoteRecord {
    private static final String USERS = "users";

    @NonNull
    private final RemoteProjectRecord mRemoteProjectRecord;

    @NonNull
    private final UserJson mUserJson;

    RemoteProjectUserRecord(boolean create, @NonNull RemoteProjectRecord remoteProjectRecord, @NonNull UserJson userJson) {
        super(create);

        mRemoteProjectRecord = remoteProjectRecord;
        mUserJson = userJson;
    }

    @NonNull
    @Override
    public UserJson getCreateObject() {
        return mUserJson;
    }

    @NonNull
    public String getId() {
        return UserData.Companion.getKey(mUserJson.getEmail());
    }

    @NonNull
    @Override
    protected String getKey() {
        return mRemoteProjectRecord.getKey() + "/" + RemoteProjectRecord.PROJECT_JSON + "/" + USERS + "/" + getId();
    }

    @NonNull
    public String getName() {
        return mUserJson.getName();
    }

    @NonNull
    public String getEmail() {
        return mUserJson.getEmail();
    }

    public void setName(@NonNull String name) {
        if (getName().equals(name))
            return;

        mUserJson.setName(name);
        addValue(getKey() + "/name", name);
    }

    public void setToken(@Nullable String token, @NonNull String uuid) {
        Assert.assertTrue(!TextUtils.isEmpty(uuid));

        Map<String, String> tokens = mUserJson.getTokens();

        if (Utils.INSTANCE.stringEquals(tokens.get(uuid), token))
            return;

        mUserJson.addToken(token, uuid);
        addValue(getKey() + "/tokens/" + uuid, token);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteProjectUserRecord.kt
package com.krystianwsul.checkme.firebase.records

import android.text.TextUtils
import com.krystianwsul.checkme.firebase.UserData
import com.krystianwsul.common.firebase.UserJson


class RemoteProjectUserRecord(
        create: Boolean,
        private val remoteProjectRecord: RemoteSharedProjectRecord,
        override val createObject: UserJson) : RemoteRecord(create) {

    companion object {

        private const val USERS = "users"
    }

    val id by lazy { UserData.getKey(createObject.email) }

    override val key by lazy { remoteProjectRecord.childKey + "/" + USERS + "/" + id }

    var name: String
        get() = createObject.name
        set(name) {
            if (name == createObject.name)
                return

            createObject.name = name
            addValue("$key/name", name)
        }

    val email by lazy { createObject.email }

    var photoUrl: String?
        get() = createObject.photoUrl
        set(value) {
            if (value == createObject.photoUrl)
                return

            check(!value.isNullOrEmpty())

            createObject.photoUrl = value
            addValue("$key/photoUrl", value)
        }

    fun setToken(token: String?, uuid: String) { // todo remove
        check(!TextUtils.isEmpty(uuid))

        if (token == createObject.tokens[uuid])
            return

        createObject.tokens[uuid] = token
        addValue("$key/tokens/$uuid", token)
    }

    override fun deleteFromParent() = check(remoteProjectRecord.remoteUserRecords.remove(id) == this)
}
